core/unix/include/sigcontext.h

```diff
@@ -352,26 +371,137 @@ struct _aarch64_ctx {
 #        define FPSIMD_MAGIC 0x46508001
 
 struct fpsimd_context {
-    struct _aarch64_ctx head;
-    __u32 fpsr;
-    __u32 fpcr;
-    __uint128_t vregs[32];
+    struct _aarch64_ctx head; /* 8 bytes */
+    __u32 fpsr;               /* 4 */
+    __u32 fpcr;               /* 4 */
+    __uint128_t vregs[32];    /* 512 */
 };
 
-/* TODO i#5365: Storage of sve_context in kernel_sigcontext_t.__reserved, see
- * above. See also sigcontext_to_mcontext_simd() and
- * mcontext_to_sigcontext_simd().
+/* Storage of sve_context in kernel_sigcontext_t.__reserved, see above. See
+ * also sigcontext_to_mcontext_simd() and mcontext_to_sigcontext_simd().
  */
 
 #        define SVE_MAGIC 0x53564501
 
 struct sve_context {
+    struct _aarch64_ctx head; /* 8 bytes */
+    __u16 vl;                 /* 2 */
+    __u16 __reserved[3];      /* 6 */
+};
+
+/*
+ * extra_context: describes extra space in the signal frame for
+ * additional structures that don't fit in sigcontext.__reserved[].
+ *
+ * Note:
+ *
+ * 1) fpsimd_context, esr_context and extra_context must be placed in
+ * sigcontext.__reserved[] if present.  They cannot be placed in the
+ * extra space.  Any other record can be placed either in the extra
+ * space or in sigcontext.__reserved[], unless otherwise specified in
+ * this file.
+ *
+ * 2) There must not be more than one extra_context.
+ *
+ * 3) If extra_context is present, it must be followed immediately in
+ * sigcontext.__reserved[] by the terminating null _aarch64_ctx.
+ *
+ * 4) The extra space to which datap points must start at the first
+ * 16-byte aligned address immediately after the terminating null
+ * _aarch64_ctx that follows the extra_context structure in
+ * __reserved[].  The extra space may overrun the end of __reserved[],
+ * as indicated by a sufficiently large value for the size field.
+ *
+ * 5) The extra space must itself be terminated with a null
+ * _aarch64_ctx.
+ */
+#        define EXTRA_MAGIC 0x45585401
+
+struct extra_context {
+    struct _aarch64_ctx head; /* 8 bytes */
+    __u64 datap; /*  8 bytes. 16-byte aligned pointer to extra space cast to __u64 */
+    __u32 size;  /* 4 bytes. size in bytes of the extra space */
+    __u32 __reserved[3]; /* 12 bytes */
+};
+
+#        define ESR_MAGIC 0x45535201
+
+struct esr_context {
     struct _aarch64_ctx head;
-    __u16 vl;
-    __u16 __reserved[3];
+    __u64 esr;
 };
+
 #    endif
 
+/* SVE helper macros. */
+#    define __SVE_VQ_BYTES 16 /* number of bytes per quadword */
```

Comment:
What is "VQ"?  Could we use a slightly longer more descriptive name?

Reply:
`VQ` means vector length in quadwords, where a quadword is 128 bits.

It turns out that almost all of these helper macros are not required in DR's `sigcontext.h` because they are available in the kernel's `sigcontext.h`.

There are no names with `VQ` or `vq` in them. `sve_vq_from_vl(vl)` has been changed to `sve_vecquad_from_veclen(veclen)`.